Fix a flakiness in test gpdispatch

huansong opened the pull request:
The test was flaky with a diff:
```
@@ -25,8 +25,10 @@

 -- session1 will be fatal.
 1: select count(*) > 0 from gp_dist_random('pg_class');
-FATAL:  could not allocate resources for segworker communication
-server closed the connection unexpectedly
+ ?column?
+----------
+ t
+(1 row)
```

The reason is that the DTX recovery process got ahead and detonate it first...
```
2023-04-20 17:03:26.380438 UTC,,,p119165,th1614203008,,,,0,con2,,seg-1,,,,sx1,"LOG","00000","DTM Started",,,,,,,0,,"cdbdtxrecovery.c”,155
...
2023-04-20 17:04:18.184150 UTC,,,p119165,th1614203008,,,,0,con2,,seg-1,,,,sx1,"LOG","XX009","fault triggered, fault name:'make_dispatch_result_error' fault type:'skip' ",,,,,,,0,,"faultinjector.c",475,
2023-04-20 17:04:18.292955 UTC,,,p119165,th1614203008,,,,0,con2,,seg-1,,,,sx1,"FATAL","XX000","could not allocate resources for segworker communication
2023-04-20 17:04:18.293370 UTC,,,p119165,th1614203008,,,,0,con132,,seg-1,,,,sx1,"LOG","00000","The previous session was reset because its gang was disconnected (session id = 2). The new session id = 132
```
...so the intended backend didn't:
```
3-04-20 17:04:18.175160 UTC,"gpadmin","isolation2test",p122404,th1614203008,"[local]",,2023-04-20 17:04:18 UTC,0,con129,cmd2,seg-1,,,,sx1,"LOG","00000","injected fault 'make_dispatch_result_error' type 'skip'",,,,,,"select gp_inject_fault('make_dispatch_result_error', 'skip', dbid) from gp_segment_configuration where role = 'p' and content = -1;",0,,"faultinjector.c",1077,
...
2023-04-20 17:04:18.187420 UTC,"gpadmin","isolation2test",p122404,th1614203008,"[local]",,2023-04-20 17:04:18 UTC,0,con129,cmd3,seg-1,,,,sx1,"LOG","00000","statement: select count(*) > 0 from gp_dist_random('pg_class');",,,,,,,0,,"postgres.c",1729
```

Let's exclude the background worker from getting that fault.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

ashwinstar commented:
Is this really one off case, or DTX recovery process similar to auto-vacuum processes can get in way for other places/fault as well. Won't like hard-line (and don't know if DTX recovery process related tests use faults or not) but maybe similar to autovacuum ban it in framework at common place? 

```
    /*                                                                                                                                      
     * Auto-vacuum worker and launcher process, may run at unpredictable times                                                              
     * while running tests. So, skip setting any faults for auto-vacuum                                                                     
     * launcher or worker. If anytime in future need to test these processes                                                                
     * using fault injector framework, this restriction needs to be lifted and                                                              
     * some other mechanism needs to be placed to avoid flaky failures.                                                                     
     */                                                                                                                                     
    if (IsAutoVacuumLauncherProcess() ||                                                                                                    
        (IsAutoVacuumWorkerProcess() &&                                                                                                     
         !(0 == strcmp("vacuum_update_dat_frozen_xid", faultName) ||                                                                        
             0 == strcmp("auto_vac_worker_before_do_autovacuum", faultName))))                                                              
        return FaultInjectorTypeNotSpecified;                                                                                               
```

huansong commented:
> Is this really one off case...

This is the first case I saw. But a quick search shows another: https://github.com/greenplum-db/gpdb/commit/8d0deb26f342a9e6d9d45ac5bf8c730e0f9bca58

> maybe similar to autovacuum ban it in framework at common place

So yes, that would be better I agree. Not sure what tests need faults in dtx process either. I'm running a pipeline to find that out... https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/flake-gpdispatch-nodtxfault

huansong commented:
>> maybe similar to autovacuum ban it in framework at common place

> So yes, that would be better I agree. Not sure what tests need faults in dtx process either. I'm running a pipeline to find that out... https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/flake-gpdispatch-nodtxfault

Only two faults that we are testing dtx recovery process with. I've modified the code to do what we did for autovacuum process. Refactored it a little bit too.

ashwinstar commented on src/backend/utils/misc/faultinjector.c:
```diff
@@ -242,6 +242,35 @@ FaultInjector_ShmemInit(void)
 }
 
+static bool
+checkBgProcessSkipFault(const char* faultName)
```
I am thinking what would be best way to make future developers and ourselves aware of this hack to avoid spending time if writing new test these background processes. Maybe we should add note to gp_inject_fault function referring to this function to check for details on which faults work and which don't for background workers.
